File: htmlFileToCsv.py
#!/usr/bin/python
from sys import argv
from os import listdir, unlink
from shutil import rmtree
from bs4 import BeautifulSoup
# Functions
Debug = False


def debug(content):
    if Debug:
        print(content)


def readfile(dir):
    return open(dir, "r").read()


def exportToFile(dir, content):
    open(dir, "w").write(content)


def appendToFile(dir, content):
    try:
        inside = open(dir, "r").read()
        open(dir, "w").write(inside + content)
    except:
        open(dir, "w").write(content)


def isNum(num):
    try:
        return float(str(num)).is_integer()
    except:
        return False


def help():
    print(__file__, "[OPTION]", "[HTML File]")
    print("This program is intended to take data from inkspace HTML files and append it to a CSV file in the same directory.")
    print("INSTRUCTIONS: save inkspace output to an HTML file in the same directory as this program. Once done, run this program. Profit.")
    print("OPTIONS:")
    print("-h | --help\t\t\tPrints this help screen.")
    print("-d | --display\t\t\tPrints the output of the CSV into the console as well as into the file.")


# Generator
def htmlToSource(file):
    lines = readfile(file).split("\n")
    #debug(("lines:", lines))
    # all the content starts after the tag <textarea reeadonly="">
    text = lines[2][lines[2].find("<textarea")+22:]
    debug(("text", text))
    # all the content makes a new line in the html, so this adds every line after <textarea>
    for i in range(3, len(lines)):
        if (text.find("</textarea>") == -1):
            text = text + "\n" + lines[i]
            debug(text)
        else:
            break  # once line has a </textarea>, it stops adding lines, but includes the line with </textarea>
    # cuts out all info after </textarea>, then splits the new dictionary to define people and their attributes
    people = text[:text.find("</textarea>")].split("\n")
    debug(("people:", people))
    return people


def getContentArray(people):
    """the purpose of the function is to take all the content in the HTML and return a matrix array.
    file: a string containing the contents from a file given by readfile().
    return: a string[][] matrix with important data scraped from file.
    """
    directory = []  # the matrix used to store conent
    for i in people:
        # gets the attribute of each individual person and defines it by a space
        attr = i.split(" ")
        debug(("attr:", attr))
        if (len(attr) >= 5):  # a person cannot have any more than 5 attributes, and if they do, they all merge into the last attribute
            #                         FirstName, LastName, timeIn, timeOut, hours, purpose
            directory.append([str(attr[0]+" "+attr[1]), attr[2], attr[3],
                              attr[4], ''.join(str(x+" ") for x in attr[5:])])
            directory[-1][4] = directory[-1][4][:len(directory[-1][4])-1]
            # Comment top and uncomment bottom if you want first and last name separate
            #directory.append([attr[0], attr[1], attr[2], attr[3], attr[4], ''.join(str(x+" ") for x in attr[5:])])
            #directory[-1][5] = directory[-1][5][:len(directory[-1][5])-1]
        elif (len(attr) > 1):  # this is added to deal with first and last name being together, in case someone only puts down their name
            hold = []
            # this combines the first and last name
            hold.append(str(attr[0]+" "+attr[1]))
            if (len(attr) > 2):  # if they did put in more than just their name, its added to attributes
                for i in attr[2:]:
                    hold.append(i)
            # combines all the attributes into an array, then return to people array
            directory.append(hold)
        else:
            directory.append(attr)
        debug(("directory:", directory))
    return directory  # returns the matrix


def arrayToCsv(directory):
    """takes a matrix and returns a string in CSV format.
    var directory: a string[][] matrix that contains the information of people at the center.
    returns: a string that contains all the information in CSV format.
    """
    cvarray = ''
    for i in range(len(directory)):
        for e in range(len(directory[i])-1):
            if (isNum(directory[i][e])):
                cvarray += (directory[i][e]+", ")
            else:
                cvarray += (directory[i][e]+", ")
        cvarray += (directory[i][-1]+"\n")
        debug(("cvarray["+str(i)+"]:", cvarray))
    return (cvarray+"\n")


# Actual Script
def main():
    input = []
    display = False  # Option to display results of CSV in console as well as names.csv
    if len(argv) > 1:
        for i in argv[1:]:
            if (i == "/help" or i == "-help" or i == "==help" or i == "-h" or i == "/h"):
                help()
                return
            elif (i == "/d"or i == "-d"or i == "/display"or i == "-display"or i == "--display"):
                display = True
            if (i.split(".")[1].lower() == "html"):
                input.append(i)
        #debug("argv: " + str(input))
    if(len(input) == 0):
        for i in listdir():
            try:
                if (i.split(".")[1] == "html"):
                    input.append(i)
            except:
                continue
        #debug("listdir: " + str(input))
    if(len(input) == 0):
        print("ERROR: No HTML Files imported and none in current directory!")
        help()
        return
    output = ''
    for i in input:
        # debug(i+":"+i.split(".")[1].lower())
        output += arrayToCsv(getContentArray(i))
        try:
            # deletes the html used as to not accidentally add it to csv again
            unlink(i)
        except:
            True
        try:  # incase the folder already doesnt exist
            rmtree(str(i.split(".")[0] + "_files"))  # removes folder
        except:
            True
    # debug(("output:",output))
    if(display):
        print(output)
    appendToFile("names.csv", output)


main()

File: htmlFileToCsv.java
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;

public class htmlFileToCsv {
    private static boolean Debug = false;

    /**
     * I leave these debug commands in case anyone wants to place them anywhere
     * inside the code to output the results
     */
    private static void debug(String content) {
        if (Debug)
            System.out.println(content);
    }

    private static void debug(String content, Boolean debug) {
        if (debug)
            System.out.println(content);
    }

    private static void debug(String[] a) {
        if (Debug)
            for (String b : a)
                System.out.println(b);
    }

    private static void debug(String[] a, Boolean debug) {
        if (debug)
            for (String b : a)
                System.out.println(b);
    }

    private static void debug(List<String> a) {
        if (Debug)
            for (String b : a)
                System.out.println(b);
    }

    private static void debug(List<String> a, Boolean debug) {
        if (debug)
            for (String b : a)
                System.out.println(b);
    }

    private static void debug(String[][] a) {
        if (Debug)
            for (String[] b : a) {
                for (String c : b)
                    System.out.print(c + " ");
                System.out.println();
            }
    }

    private static void debug(String[][] a, Boolean debug) {
        if (debug)
            for (String[] b : a) {
                for (String c : b)
                    System.out.print(c + " ");
                System.out.println();
            }
    }

    /*
     * private static void debug(List<String[]> a) { if (Debug) for (String[] b : a)
     * { for (String c : b) System.out.print(c + " "); System.out.println(); } }
     * 
     * private static void debug(List<String[]> a, Boolean debug) { if (debug) for
     * (String[] b : a) { for (String c : b) System.out.print(c + " ");
     * System.out.println(); } }
     */

    /**
     * This command will take a string name for a file and attempt to output the
     * files contents in the form of a String array
     * 
     * @param file is the name of the file to read
     * @return is a String[] that contains each line in the file
     * 
     * @author Nick Cottrell
     * @version 1.0
     * @since 2019-03-08
     */
    private static String[] readFile(String file) {
        List<String> result = new ArrayList<String>();
        if (file.endsWith(".html")) {
            File html = new File(file);
            try {
                Scanner scfile = new Scanner(html);
                while (scfile.hasNextLine()) {
                    result.add(scfile.nextLine());
                }
                scfile.close();
                String[] fin = new String[result.size()];
                fin = result.toArray(fin);
                return fin;
            } catch (FileNotFoundException e) {
                return new String[1];
            }
        }
        return new String[1];
    }

    /**
     * This is the function that runs when no html files are added to the programs
     * arguements. it will search its current directory for html files to act on and
     * use them instead.
     * 
     * @return is a String[] which contains the names of all html files in current
     *         directory
     */
    private static String[] getFilesInDirectory() {
        int count = 0;
        File[] files = new File(".").listFiles();
        for (File a : files)
            if (a.getName().endsWith(".html"))
                count += 1;
        String[] names = new String[count];
        count = 0;
        for (int f = 0; f < files.length; f++) {
            if (files[f].getName().endsWith(".html")) {
                names[f] = files[count].getName();
                count += 1;
            }
        }
        return names;
    }

    /**
     * This function takes the lines given from a file and searches for specific
     * data that relates to peoples sign ins. it then converts the data into a data
     * matrix with each row specifying a person, and each column defining an
     * attribute of each person.
     * 
     * @param htmlContent the lines given by the file
     * @return a String[][] with each persons attributes
     * 
     * @author Nick COttrell
     * @version 1.0
     * @since 2019-03-08
     */
    public static String[][] getContentToArray(String[] htmlContent) {
        String text = htmlContent[2].substring(htmlContent[2].indexOf("<textarea") + 22);
        debug("Text: " + text);
        for (int i = 3; i < htmlContent.length; i++) {
            if (htmlContent[i].indexOf("</textarea>") == -1) {
                text += "\n" + htmlContent[i];
                // debug(text);
            } else { // once line has a </textarea>, it stops adding lines, but includes the line
                     // with </textarea>
                text += "\n" + htmlContent[i].substring(0, htmlContent[i].indexOf("</textarea>"));
                break;
            }
        }
        debug(text);

        /* Dictionary Creation */
        List<String[]> directory = new ArrayList<String[]>(); // the matrix that will be used to store content
        String[] people = text.split("\n");
        String[] attr;
        String[] hold = new String[5]; // there will only every be 5 attributes for anyone to have ever. that will only
                                       // change if we ever make first and last name seperate. in that case, wed need a
                                       // new function to account for that
        String purpose = "";
        for (String i : people) {
            attr = i.split(" ");
            purpose = "";
            if (attr.length >= 5) {
                for (int p = 5; p < attr.length - 1; p++)
                    purpose += attr[p] + " ";
                purpose += attr[attr.length - 1];
                hold = new String[] { (attr[0] + " " + attr[1]), attr[2], attr[3], attr[4], purpose };
                directory.add(hold);
            } else if (attr.length >= 1) {
                hold = new String[5];
                if (attr.length > 1) {
                    hold[0] = (attr[0] + " " + attr[1]);
                    for (int a = 2; a < attr.length; a++) {
                        hold[a - 1] = attr[a];
                    }
                } else {
                    hold[0] = attr[0];
                }
                for (int b = attr.length; b < hold.length; b++) {
                    hold[b] = " "; // since all the spots in the matrix length and width wise, i had to make the
                                   // attributes equals, no matter what. to combat this, i fill all empty slots
                                   // with spaces to be dealt with when converting to csv.
                }
                directory.add(hold);
            }
        }

        /* packing dictionary into matrix */
        String[][] result = new String[directory.size()][5];
        result = directory.toArray(result);

        /* double checking array to remove null values */
        for (int row = 0; row < result.length; row++)
            for (int col = 0; col < result[row].length; col++)
                if (result[row][col] == null) // if the value is outright null (could never figure out how), it will
                                              // replace it with a space.
                    result[row][col] = " ";
        return result;
    }

    /**
     * This function is depricated for a better output option. takes in a dictionary
     * matrix and converts it into a csv friendly string
     * 
     * @deprecated replaced by arrayToCsvList()
     * @param directroy is the matrix containing everyones information.
     * @return everyones information inside a string in csv format.
     * @see {@link htmlFileToCsv#arrayToCsvList(String[][])}
     * 
     * @author Nick COttrell
     * @version 1.0
     * @since 2019-03-08
     */
    public static String arrayToCsvString(String[][] directory) {
        String cvArray = "";
        for (String[] row : directory) {
            cvArray += row[0];
            for (int col = 1; col < row.length; col++) {
                debug("row[col]: " + row[col]);
                if (!row[col].equals(" "))
                    cvArray += ", " + row[col];
            }
            cvArray += "\n";
        }
        return cvArray + "\n";
    }

    /**
     * This function will take the database of peoples information and make it into
     * a string list in csv format to be used by an output function to a file
     * 
     * @param directory is a String[][] matrix that contains everyones attributes
     * @return a List<String> that containes each individual line in a csv format
     * 
     * @author Nick Cottrell
     * @version 1.0
     * @since 2019-03-08
     */
    public static List<String> arrayToCsvList(String[][] directory) {
        List<String> result = new ArrayList<String>();
        String cvArray = ""; // this line is used to add and organize all of the files contents
        for (String[] row : directory) {
            cvArray = row[0]; // immediately adds date/name. next line will add ", " so that it is never at
                              // the end of a line
            for (int col = 1; col < row.length; col++) {
                debug("row[col]: " + row[col], false);
                if (!row[col].equals(" ")) // if any values were null/empty, it will not categorize them into the file
                    cvArray += ", " + row[col];
            }
            result.add(cvArray);
        }
        return result;
    }

    public static void main(String args[]) {
        List<String> input = new ArrayList<String>();
        Boolean reveal = false; // remove when done with experiment
        if (args.length > 0) {
            for (String i : args) {
                if (i.endsWith(".html"))
                    input.add(i);
                else if ((new File(i + ".html")).exists())
                    input.add(i + ".html");
                else if (i.equals("/reveal"))
                    reveal = true;
            }
        } else {
            for (String i : getFilesInDirectory()) {
                if (i.endsWith(".html"))
                    input.add(i);
            }
        }
        debug(input);
        List<String> output = new ArrayList<String>(); // this is roughly the entire object since it holds all the
                                                       // information
        String[][] Database;
        File trash;
        for (String i : input) {
            output = arrayToCsvList(getContentToArray(readFile(i)));
            output.add(""); // this line helps separate multiple file scrapings for better organization
            Database = getContentToArray(readFile(i)); // this is the object that contains all the information on each
                                                       // person
            /* removing html file and baggage */
            trash = new File(i);
            if (trash.isDirectory()) {
                for (File e : trash.listFiles())
                    e.delete();
            }
            trash.delete();
            /*
             * Removing folder that normally goes along with html file, in fashion
             * %htmlFile%_files
             */
            String name = trash.getName();
            trash = new File(name.substring(0, name.indexOf(".html")) + "_files");
            if (trash.isDirectory()) {
                for (File e : trash.listFiles())
                    e.delete();
            }
            trash.delete();
        }
        debug(output, false);

        if (reveal) {
            debug(output, true);
        }
        /* outputing to file */
        try {
            Files.write(Paths.get("names.csv"), output, StandardCharsets.UTF_8, StandardOpenOption.APPEND);
        } catch (IOException e) {
            e.printStackTrace();
        }

    }
}